components/script/dom/writablestreamdefaultwriter.rs

```diff
@@ -0,0 +1,490 @@
+        self.closed_promise.borrow().set_promise_is_handled();
+    }
+
+    pub(crate) fn set_ready_promise(&self, promise: Rc<Promise>) {
+        *self.ready_promise.borrow_mut() = promise;
+    }
+
+    pub(crate) fn resolve_ready_promise_with_undefined(&self) {
+        self.ready_promise.borrow().resolve_native(&());
+    }
+
+    pub(crate) fn resolve_closed_promise_with_undefined(&self) {
+        self.closed_promise.borrow().resolve_native(&());
+    }
+
+    /// <https://streams.spec.whatwg.org/#writable-stream-default-writer-ensure-ready-promise-rejected>
+    pub(crate) fn ensure_ready_promise_rejected(
+        &self,
+        global: &GlobalScope,
+        error: SafeHandleValue,
+        can_gc: CanGc,
+    ) {
+        let mut ready_promise = self.ready_promise.borrow_mut();
+        // If writer.[[readyPromise]].[[PromiseState]] is "pending", reject writer.[[readyPromise]] with error.
+        if ready_promise.is_pending() {
+            ready_promise.reject_native(&error);
+        } else {
+            // Otherwise, set writer.[[readyPromise]] to a promise rejected with error.
+            let promise = Promise::new(global, can_gc);
+            promise.reject_native(&error);
+            *ready_promise = promise;
+        }
+        // Set writer.[[readyPromise]].[[PromiseIsHandled]] to true.
+        ready_promise.set_promise_is_handled();
+    }
+
+    /// <https://streams.spec.whatwg.org/#writable-stream-default-writer-ensure-closed-promise-rejected>
+    pub(crate) fn ensure_closed_promise_rejected(
+        &self,
+        global: &GlobalScope,
+        error: SafeHandleValue,
+        can_gc: CanGc,
+    ) {
+        let mut closed_promise = self.closed_promise.borrow_mut();
+        // If writer.[[closedPromise]].[[PromiseState]] is "pending", reject writer.[[closedPromise]] with error.
+        if closed_promise.is_pending() {
+            closed_promise.reject_native(&error);
+        } else {
+            // Otherwise, set writer.[[closedPromise]] to a promise rejected with error.
+            let promise = Promise::new(global, can_gc);
+            promise.reject_native(&error);
+            *closed_promise = promise;
+        }
+        // Set writer.[[closedPromise]].[[PromiseIsHandled]] to true.
+        closed_promise.set_promise_is_handled();
+    }
+
+    /// <https://streams.spec.whatwg.org/#writable-stream-default-writer-abort>
+    fn abort(
+        &self,
+        cx: SafeJSContext,
+        global: &GlobalScope,
+        reason: SafeHandleValue,
+        can_gc: CanGc,
+    ) -> Rc<Promise> {
+        // Let stream be writer.[[stream]].
+        let Some(stream) = self.stream.get() else {
+            // Assert: stream is not undefined.
+            unreachable!("Stream should be set.");
+        };
+
+        // Return ! WritableStreamAbort(stream, reason).
+        rooted!(in(*cx) let mut reason_clone = UndefinedValue());
+        reason_clone.set(reason.get());
```

Comment:
This should be unnecessary with my suggestion for abort.

Reply:
Done